contracts/src/v0.8/llo-feeds/v0.5.0/configuration/Configurator.sol

```diff
@@ -33,44 +35,152 @@ contract Configurator is IConfigurator, ConfirmedOwner, TypeAndVersionInterface,
   error InsufficientSigners(uint256 numSigners, uint256 minSigners);
 
   struct ConfigurationState {
-    // The number of times a new configuration
-    // has been set
+    // The number of times a configuration (either staging or production) has
+    // been set for this configID
     uint64 configCount;
     // The block number of the block the last time
-    /// the configuration was updated.
+    // the configuration was updated.
     uint32 latestConfigBlockNumber;
+    // isGreenProduction is a bit flip that indicates whether blue is production
+    // exactly one of blue/green must be production at all times.
+    // 0 -> blue is production
+    // 1 -> green is production
+    //
+    // So, to clarify, if isGreenProduction is false (initial state) then:
+    // [0](blue) is production and [1](green) is staging/retired
+    //
+    // and if isGreenProduction is true then:
+    // [0](blue) is staging/retired and [1](green) is production
+    // State is swapped every time a staging config is promoted to production.
+    bool isGreenProduction;
+    // The digest of the current configurations (0 is always blue, 1 is always green)
+    bytes32[2] configDigest;
   }
 
   constructor() ConfirmedOwner(msg.sender) {}
 
-  /// @notice Configuration states keyed on DON ID
+  /// @notice Configuration states keyed on config ID
+  /// @dev The first element is the blue configuration state
+  /// and the second element is the green configuration state
   mapping(bytes32 => ConfigurationState) internal s_configurationStates;
 
   /// @inheritdoc IConfigurator
-  function setConfig(
-    bytes32 donId,
-    address[] memory signers,
+  function setProductionConfig(
+    bytes32 configId,
+    bytes[] memory signers,
     bytes32[] memory offchainTransmitters,
     uint8 f,
     bytes memory onchainConfig,
     uint64 offchainConfigVersion,
     bytes memory offchainConfig
   ) external override checkConfigValid(signers.length, f) onlyOwner {
+    require(onchainConfig.length == 64, "Invalid onchainConfig length");
+
+    // Ensure that predecessorConfigDigest is unset and version is correct
+    uint256 version;
+    bytes32 predecessorConfigDigest;
+    assembly {
+      version := mload(add(onchainConfig, 32))
+      predecessorConfigDigest := mload(add(onchainConfig, 64))
+    }
+    require(version == SUPPORTED_ONCHAIN_CONFIG_VERSION, "Unsupported onchainConfig version");
+    require(predecessorConfigDigest == 0, "predecessorConfigDigest must be unset for production config");
+
     _setConfig(
-      donId,
+      configId,
       block.chainid,
       address(this),
       signers,
       offchainTransmitters,
       f,
       onchainConfig,
       offchainConfigVersion,
-      offchainConfig
+      offchainConfig,
+      true
+    );
+  }
+
+  /// @inheritdoc IConfigurator
+  function setStagingConfig(
```

Comment:
The person setting config only cares about production vs. staging. They don't need to know (and it could be difficult to find out) which of blue/green is production at any given time.